### ISSUE TITLE
Feature: Use makefile output in configurations

### ISSUE THREAD

**Issue by KrestenHelstrup** [microsoft/vscode-makefile-tools#87]
I don't know if this is possible at all but it would be very handy to use output from the makefile in my configurations. I'm trying to setup launch and debug configurations and as far as I can see I have type in the name of my .bin and .out files but in principle they are already known by analyzing the makefile.
Let me know what you think

**Comment by andreeis** [comment]
This should work already. What do you see when you click on the edit pen near "launch configurations" in the Makefile Tools left side panel?

To just write the theory of it:
- the extension looks at the output of a live dryrun (or the user provided text log) and when it identifies a linker command, searches for -o switch and grabs the file name
- if the -o is missing, it deduces the file name according to some rules, this applies mostly on windows cl/link since on others if -o is missing I know of just the a.out rule
- if we think it's an object or a library most of the times we don't include that in the list
- then, more than what the linking provides, if we identify any binary that was built with a link command and which is called during the makefile execution, we take the folder from where it appears to be run and the arguments given and we add that to the list

Then, once you chose one such launch configuration in the UI side panel, you can invoke them via the commands makefile.launchDebug or makefile.launchRun, which also have buttons in the UI side panel.

If you want to debug via F5, that will use a launch.json and you can still inject Makefile Tools knowledge in that like below.

"cwd": "${command:Makefile.launchTargetDirectory}",
"name": "Launch Program",
"program": "${command:Makefile.launchTargetPath}",
"args": ["${command:Makefile.launchTargetArgs}"]

Sometimes you may need since VSCode (not the exension) has some issues with the syntax for string array for args, I forgot right now what is exactly but when you encounter the issue it's obvious that the above won't work and you need the below. launchTargetArgsConcat will send one string with all the arguments concatenated and separated by space (no special quoting yet).
"args": ["${command:Makefile.launchTargetArgsConcat}"]

**Comment by KrestenHelstrup** [comment]
At the moment it looks like this when I press the edit pen:
![launchTargets](https://user-images.githubusercontent.com/74759135/105720294-90a29400-5f23-11eb-8917-c25b770076f8.PNG)
And my launch config looks like this - Here you see the VsCodeTestSoftware.bin that I would like to get rid of.
![LaunchConfig](https://user-images.githubusercontent.com/74759135/105720309-939d8480-5f23-11eb-946a-3b2b265331b3.PNG)
This is needed because it is not a windows program I'm developing, it is an embedded program, so I need another application to program the binary into the target. 
Where do I find a list of variables available ? I have not seen any documentation with it and auto-completion doesn't seem to work?

**Comment by andreeis** [comment]
We are working on documentation. You should see auto-completion for settings in settings.json. For variables in launch.json it doesn't work yet, right.

We will think how to better support the embedded scenarios for debugging. I didn't play with that yet. Your collaboration would help much with this. 
In the meantime, you can simply edit the makefile.launchConfigurations section as you need. Let me know how it works after you edit.

**Comment by KrestenHelstrup** [comment]
Maybe I just don't get it - But should I be able to replace the "armgcc/exe/VsCodeTestSoftware.bin" with any of those variables ?
At least that is not working and when I try to launch I get the following output:
"..\..\..\..\..\..\ARMTC\Simplicity Commander\Commander.exe" flash ${command:Makefile.launchTargetArgsConcat} --device efm32g890f128 --force
So it looks like BinaryArgs does not try to translate the variable at all.
Further I also have to specify --device XXX which I also have in the makeFile like -DDEVICE=EFM32G890F128 - Is it possible to use that in this context ?

**Comment by andreeis** [comment]
I forgot to mention, we didn't implement yet the variable expansion in settings.json. Use real paths (absolute or relative to the workspace root - no the .vscode subfolder -).

I need more details about the DEVICE define. Share please the fragment of dryrun.log that shows how -DDEVICE is used. Or better share he whole dryrun.log. The one you shared in issue 88 was having only compilation/linking commands, I saw no DEVICE define in there.

I can't wait to see your project end to end, if you'll ever have time to put up a project on GitHub to help with our testing (and to better and faster address your reports).

**Comment by KrestenHelstrup** [comment]
Sorry - I have a mistake in the makefile. I have attached a new dryrun. 
I don't know what you mean about sharing only a part of it? I have shared the entire file.
[dryrun.log](https://github.com/microsoft/vscode-makefile-tools/files/5871694/dryrun.log)

I will try to put the project on GitHub in the weekend - I see know that it will be much easier to show what I want to achieve.
The basic idea is that I should have all the information in the makefile so I should be able to get the information from there so I don't have to type it in multiple configs. As it is today, if I want to change the name of the project (version update) I have to update in makefile, settings.json (to get run working) and launch.json (to get debug working). 
The new dryrun shows what the version numbers could like look.

**Comment by andreeis** [comment]
Entire file is best, some users may opt to share fragments due to IP issues (although you were willing to put on GitHub so I didn't need to mention).

To get run working you can still use launch.json instead of settings.json. Run the command "Debug: Start without debugging" which will read launch.json (same as F5 does for debugging). If you have CMake Tools activated on your project, run that command from the palette, not via the default key bindings, since it overlaps with "CMake: Debug". So 2 updates instead of 3.

Now, back to you having to update only one place. Yes, we try to deduce as much as possible from the makefile. In this case, by looking at the latest dryrun and what you're trying to achieve, I don't think we can, but let me explain because you may still be able to change your makefile so that the extension will understand.

The extension will add one entry (with cwd, binary and args) into makefile.launchConfigurations if:
- it finds a linker command that produces a -o output that is binary (not object, not library), with some limitations where it deduces that wrong. In this case, args is an empty array because just from a link command we don't know how the user wants to invoke that binary later. But we have CWD (the place where it got built) and binary name.
- from the list of any binaries identified via the above situation, it also identifies if the makefile is ever invoking it throughout the execution of the makefile. It must have the same CWD (so we're not tracking yet renames, removes and copies). This may produce some arguments, depending what the makefile output (whether a live dryrun or a given build log) contains.

In your case, there are some reasons for the extension not being able to guess what's needed:
- the DEVICE define appears only for compilation commands. We can't know that you'd ever need that when you invoke a binary target
- you do a rename from .out to .bin, we will track that in the future
- the dryrun contains no info about commander.exe, we can't know you need that. We will only look at binaries that are built. Otherwise we may end up for other repos producing unreasonable big list of launch targets, every time the makefile runs a tool of any kind.

Brainstorming what you can do now to get a better experience in the debug/launch space:
- focus on launch.json, debug with F5 or the command "Debug: Start Debugging", run with Ctrl+F5 (if you don't have CMake activated) or the command "Debug: Start without debugging"
- avoid the copy from .out to .bin, link directly into .bin, is that possible? This way we will be able to identify automatically launchTargetPath. If you really need both, copy from .bin into .out, the other way around.
- use smartly the launchTargetPah and launchTargetArgs(or with Concat suffix) like this:
"program": "commander.exe", (as opposed to the real binary which is the most common scenario)
"args": ["flash ${command:Makefile.launchTargetPath} ${command:Makefile.launchTargetArgs}"]
But you need to create a new dummy target in your makefile that would appear to invoke your arm binary (even if it won't work, it won't hurt, but dryrun will see it and we'll parse it):
MyDummyTarget: VsCodeTestSoftware_1-0.bin --device $(DEVICE) --force
Make sure MyDummyTarget will be hooked up in the whatever main target is run in general for your project (whether "all" or the empty string default).

I expect this way you'll be able to change the DEVICE define in one place.

**Comment by andreeis** [comment]
I am not sure what I suggested last would even work. VSCode will try to debug the wrapper commander.exe. I have no idea if you'll ever hit breakpoints in the sources of VsCodeTestSoftware_1-0.bin.
Did you try that already (even if you write all manually in 3 places)?

**Comment by KrestenHelstrup** [comment]
I will try what you posted.

I don't have the cmake extension since I use make - Would I benefit from it ?

From your post it sounds like you might be mixing running and debugging. I have launch.json to run a debug session and makefile.launchconfiguration in settings.json to program and run the software without debugging. I have attached both files to you can see the difference. For both of the cases I need to update either "executable" or "binaryArgs" if the name of the bin/out file changes.
[launch.json.txt](https://github.com/microsoft/vscode-makefile-tools/files/5872428/launch.json.txt)
[settings.json.txt](https://github.com/microsoft/vscode-makefile-tools/files/5872427/settings.json.txt)

**Comment by andreeis** [comment]
No need for CMake extension. I only wanted to make sure that by using Ctrl+F5 you won't get "CMake: Debug" instead of "Debug: Start without debugging".

I admit I don't understand completely the complexity of the embedded debugging/running. Maybe I over simplified. So in the end you need both .bin and .out, but the extension will be able to automatically detect as a binary target only one, whatever the link command produces. Not sure what to do here, I'll think more.

Then I also don't understand your launch.json, does that work at all? For example, I know of "program" attribute, not of "executable". And I never saw the "device" attribute before and others. 

To avoid settings.json (only until we implement the variable expansion), in launch.json you can define 2 configurations, one for run and one for debug. Just a thought. You can toggle which one of the two from a status bar button on bottom-left. Let me know if you don't see it.

Can you attach the extension.log after configuring with your latest makefile change? The equivalent of the last dry-run that you attached. I want to see which launch target is identified by the extension: the .bin or the .out. From the dry-run I would see that it finds .out but from your settings and launch.json it looks like the extension finds the .bin.

Anoher idea to work around the .bin versus .out thing. Link into VsCodeTestSoftware (no exension). Then copy .bin and .out, continue everything else the same. This will cause makefile.launchTargetPath to be VsCodeTestSoftware and you can edit manually the extension as a suffix after the variable.
And you create 2 configurations in launch.json, one for debug and one for run (I see now how they are different). And still do one edit.

**Comment by andreeis** [comment]
We can also support a new command launchTargetPathBase or something (that would return only the file name) so that you don't do the trick with removing the extension. But until then, you can do what I suggested in the last paragraph of my above reply.

**Comment by KrestenHelstrup** [comment]
The launch.json is utilizing an extension called cortex-debug - and yes it is working perfectly fine.
See: [extension.log](https://github.com/microsoft/vscode-makefile-tools/files/5872848/extension.log)

I cant seem to get the ${command:Makefile.xxx} variables working - see picture below:
![CommandIssues](https://user-images.githubusercontent.com/74759135/105834579-f6962680-5fca-11eb-8281-1ad40346c757.PNG)
Do I need to do any special things to get it working ?

Would be lovely to have launchTargetPathBase - That would allow me to do the daily/weekly version changes in makefile without updating any of the configuration files. Should also be quite easy for you to supply if you already have it with extension :)

If I want to make two launch configurations - one for run and one for debug - what would I put in type of the run ?
I only need to execute an external exe file with specific args.

I'm starting to think that I will upload a "getting started" project to GitHub that will help others getting embedded development running with vscode-makefile-tools.

**Comment by andreeis** [comment]
My bad, lower case M.
Yes, we'll do ... lets' say ... launchTargetFilenameBase and you'll also use launchTargetDirectory before that, so that we don't include the directory in this new command, to let it be only the file name.

As type of the run I think the same, I know it's confusing to have a debug wording for launch, but from what I saw, the regular Debug extension at least, is able to run 2 different commands (debug and run without debugging) on the same launch.json configuration. Maybe cortex debug is different, I don't know.

Thank you very much for your help and collaboration. And a "getting started" project for embedded development would be awesome!

**Comment by KrestenHelstrup** [comment]
The commands work now with the lower case m.

What is the expected timeframe, if at all possible to say, regarding addition of variable expansion in settings.json ?

Now I'm just thinking out loud, but would it be possible to add something like ${command:makefile.makefileDefines[xxx]} with an array of all defines detected by your extension - So I could do ${command:makefile.makefileDefines[DEVICE]} which would expand to EFM32G890F128.

Maybe I'm going out of scope of the ticket - but is it possible to add a key bind to makefile build, clean or run on a workspace basis? So i.e. added to settings.json ? 
At the moment I have a hard time figuring out how to do these things in the easiest way.

Thanks for all your help also.

**Comment by andreeis** [comment]
Variables expansion: not sure but at least a few months from now.
Interesting ideas, let us think about them.
If anything feels more complicated than it should be let us know. We may not fix it for the next release which we hope to finalize soon (we're wrapping up) but for sure we will consider in the next release. We will also release a blog post and some documentation in the repository.
Are you using the Makefile Tools side panel on the left? It has buttons and context menus that run the commands in the pallette.

**Comment by KrestenHelstrup** [comment]
Yes, I use the side panel - But it seems like VSCode encourages users to do as much as possible with the keyboard only.
Also coming from eclipse I'm used to build with ctrl+b.

Having an easy way to do this is especially important when debugging via F5 does not force a rebuild.

**Comment by andreeis** [comment]
Implementing "save before build" and "build before debug" now, for release 0.1.4.
Will fork later new GitHub reports from the suggestions found throughout this thread.
And you can define any keyboard shortcuts you want for the commands in the pallette.

**Comment by KrestenHelstrup** [comment]
Thanks for the updates :)
Any expected date for 0.1.4 ?

**Comment by andreeis** [comment]
If all is well we're thinking middle next week. You shouldn't need to wait for it though. You can apply the latest vsix built before the last PR got merged in. Use "Install from VSIX" feature of VSCode.
https://github.com/microsoft/vscode-makefile-tools/actions/runs/533974103#artifacts

Once the new settings are in place, I'll run the "Build VSIX" action again.

**Comment by KrestenHelstrup** [comment]
I tried it today. I don't think the "build before debug" works as intended. At least if I press f5 it only triggers a save but not a new build. 
Easiest way to reproduce/test seems to be deleting the executable needed by debug, press f5, see error about missing file.

Pressing launch or debug from the makefile-tools panel does trigger a rebuild, but for embedded programming I need another extensions to take care of the debugging (which is launched via f5). I don't know if it is possible, and maybe I already asked, but it would be quite nice if all of the different debug buttons could point to the same action.

Save before build works as expected - Thanks.

**Comment by andreeis** [comment]
Thank you for trying the latest vsix. I forgot about hooking up "build before debug" with the F5 scenario. Of course. Fixing that now and will have a new vsix to try soon.
We plan to revisit at some point how all these debug/run scenarios fit together. The same design applies to CMake Tools extension as well and there is some history involved in why things are so now.

**Comment by andreeis** [comment]
Here's a new vsix that you can try: https://github.com/microsoft/vscode-makefile-tools/actions/runs/556765072.
We renamed the launch variables: launchTargetPath --> getLaunchTargetPath. And all others have a get prefix now. These variables work as before and they don't trigger a build.
We added 2 new launch commands launchTargetPath and launchTargetFileName (the names that were before) which also trigger a build when used. Throughout your launch.json, you should invoke only one of these new variables and only once. I know you need more invocations of these for your scenario, use the "get" prefixed versions for all except one. Otherwise, F5 will spawn more than one build.
We will implement soon support for tasks and that will offer you a better way to hook build before launch, via preLaunchTask in launch.json. At that time, we will deprecate launchTargetPath and launchTargetFileName.
